tests/dataflow/test_unified_systolic.py

```diff
@@ -6,112 +6,151 @@
 import allo.dataflow as df
 import allo.backend.hls as hls
 import numpy as np
+import pytest
 
 M, N, K = 4, 4, 4
 Rt, Ct = 2, 2
-# Rt, Ct = 4, 4
-# M, N, K = 16, 16, 16
-# Rt, Ct = 4, 4
-# Rt, Ct = 16, 16
 P0, P1 = Rt + 2, Ct + 2
 
 
 @df.region()
 def top():
-    fifo_R = df.array(df.pipe(dtype=int32, shape=(), depth=4), shape=(P0, P1))
-    fifo_C = df.array(df.pipe(dtype=int32, shape=(), depth=4), shape=(P0, P1))
+    # interconnect
+    fifo_R = df.array(df.pipe(dtype=int32, shape=(), depth=16), shape=(P0, P1 - 1))
+    fifo_C = df.array(df.pipe(dtype=int32, shape=(), depth=16), shape=(P0 - 1, P1))
+    inst_broad = df.array(df.pipe(dtype=bool, shape=(), depth=4), shape=(P1 - 1,))
+    inst_chain = df.array(df.pipe(dtype=bool, shape=(), depth=4), shape=(P0 - 1, P1))
 
     @df.kernel(mapping=[P0, P1])
-    def gemm(A: int32[M, K], B: int32[K, N], flowtag: bool, C: int32[M, N]):
+    def gemm(A: int32[M, K], B: int32[K, N], inst: bool[1], C: int32[M, N]):
 
-        # --------------------------------------------------------
-        Rtimes: int32 = M // Rt if flowtag else K // Rt
-        Ctimes: int32 = N // Ct
-        Tlength: int32 = K if flowtag else M
-        local_S: int32
         i, j = df.get_pid()
 
+        # --------------------------------------------------------
+        # Decode and Dispatch
+
+        with allo.meta_if(i == 0 and j == 0):
+            tag: bool = inst[0]
+            inst_broad[j].put(tag)
+            inst_chain[i, j].put(tag)
+
+        with allo.meta_else():
+            with allo.meta_if(i == 0):
+                flowtag: bool = inst_broad[j - 1].get()
+            with allo.meta_else():
+                flowtag: bool = inst_chain[i - 1, j].get()
+
+            with allo.meta_if(i == 0 and j != P1 - 1):
+                inst_broad[j].put(flowtag)
```

Comment:
It seems there will be both read and write in the same kernel?

Reply:
Yeah, but the read and write operation are on different fifo, so I think that's not a problem.

Reply:
There're no warnings during csim?

Reply:
No, the csim is purely clean with no relevant warning.